tests: ensure that pre-submits get additional reviews

pohly opened the pull request:
Blocking pre-submit jobs must be for stable, important features. Non-blocking pre-submit jobs should only be run automatically if they meet the criteria outlined in https://github.com/kubernetes/community/pull/8196.

To ensure that this is considered when defining pre-submit jobs, they need to be listed in `config/tests/jobs/presubmit-jobs.yaml`. UPDATE_FIXTURE_DATA=true re-generates that file to the expected state for inclusion in a PR.

aojea commented on config/tests/jobs/presubmit-jobs.yaml:
```diff
@@ -0,0 +1,156 @@
+always_run:
```
we use to put a header to indicate that is autogenerated and should not be modified manually 

> # AUTOGENERATED by ...  DO NOT EDIT!

pohly replied:
Added.

aojea commented:
/assign @aojea @BenTheElder 

seems an interesting approach that may solve the existing problem, @BenTheElder is more familiar with all the edges here, let's hear his thoughts

pohly commented:
https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/test-infra/33958/pull-test-infra-unit-test/1868608463559462912

The test worked, my branch was out-dated and thus I didn't include an up-to-date generated file:
```diff
$ git diff
diff --git a/config/tests/jobs/presubmit-jobs.yaml b/config/tests/jobs/presubmit-jobs.yaml
index b7415779ba..197152ed12 100644
--- a/config/tests/jobs/presubmit-jobs.yaml
+++ b/config/tests/jobs/presubmit-jobs.yaml
@@ -25,7 +25,7 @@ run_if_changed:
       run_if_changed: ^(go.mod|go.sum|vendor)
     - name: pull-kubernetes-apidiff-client-go
       optional: true
-      run_if_changed: (^staging\/src\/k8s.io\/client-go)|(^staging\/src\/k8s.io\/code-generator\/examples)
+      run_if_changed: ((^staging\/src\/k8s.io\/client-go)|(^staging\/src\/k8s.io\/code-generator\/examples))/.*\.go
     - name: pull-kubernetes-conformance-image-test
       optional: true
       run_if_changed: conformance
```

pohly commented on config/tests/jobs/presubmit-jobs.yaml:
```diff
@@ -0,0 +1,158 @@
+      run_if_changed: /staging/src/k8s.io/component-base/logs/
+    - name: pull-kubernetes-kind-text-logging
+      optional: true
+      run_if_changed: /staging/src/k8s.io/component-base/logs/
```
These two may qualify for merge-blocking now. I'll check...

Having them show up in this list is a useful reminder :sweat_smile:

pohly replied:
Apropos "merge-blocking": as discussed on Slack, we don't want jobs to run only occasionally *and* be merge-blocking. We currently have one such job. I demoted it to `optional: true`, changed the test so that it errors when it encounters such a job (that's how I found it), and removed the `optional` from the file because all `run_if_changed` jobs now will be optional.

pohly replied:
Looks like I didn't? Still seems like a good idea, so I'll add it.

aojea commented:
LGTM

tagging @michelle192837 too

aojea commented on config/tests/jobs/jobs_test.go:
```diff
@@ -1393,3 +1396,180 @@ func TestKubernetesProwJobsShouldNotUseDeprecatedScenarios(t *testing.T) {
+		return false
+	})
+
+	// If a job has the same settings regardless of the branch, then
+	// we can reduce to a single entry without the branch info.
+	shorterJobs := make([]presubmitJob, 0, len(*jobs))
```
any  possible side effect of this simplification?

pohly replied:
The job still shows up as intended in `config/tests/jobs/presubmit-jobs.yaml` with all of it's settings, we just don't show for which branches it gets defined. For that a reviewer would have to look at the other YAMLs. Forking a job for a release does not change `config/tests/jobs/presubmit-jobs.yaml` which IMHO is desirable.

However, it opens a small loophole: a job could get defined and accepted for master, then later someone adds variants for release branches without that triggering a detailed review because `config/tests/jobs/presubmit-jobs.yaml` doesn't get touched. I think this is acceptable.

aojea replied:
> with all of it's settings, we just don't show for which branches it gets defined. For that a reviewer would have to look at the other YAMLs.

is it possible to add a comment that shows the versions 
```
# also available in v1.31, v1.32
```
or is that not worth it?

pohly replied:
Reducing `Branches` and `SkipBranches` (a regular expression!) to simple version numbers would be a bit challenging, albeit solvable.

But the bigger question is: do we *want* this file to change when adding the same job to a release branch? It closes a loophole, but also adds churn during each release.

pohly replied:
@BenTheElder: thoughts on this?

BenTheElder replied:
I think it is a good trade-off to ignore the branch regexes for this purpose.

aojea commented on config/tests/jobs/presubmit-jobs.yaml:
```diff
@@ -0,0 +1,158 @@
+    - name: pull-kubernetes-conformance-image-test
+      optional: true
+      run_if_changed: conformance
+    - name: pull-kubernetes-conformance-kind-ipv6-parallel
```
this can be removed

aojea replied:
done in https://github.com/kubernetes/test-infra/pull/33965

aojea commented on config/tests/jobs/presubmit-jobs.yaml:
```diff
@@ -0,0 +1,158 @@
+    - name: pull-kubernetes-e2e-autoscaling-hpa-cpu
+      optional: true
+      run_if_changed: ^(pkg\/controller\/podautoscaler\/|test\/e2e\/autoscaling\/horizontal_pod_autoscaling|test\/e2e\/framework\/autoscaling\/)
+    - name: pull-kubernetes-e2e-capz-azure-disk
+      optional: true
+      run_if_changed: azure.*\.go
+    - name: pull-kubernetes-e2e-capz-azure-disk-vmss
+      optional: true
+      run_if_changed: azure.*\.go
+    - name: pull-kubernetes-e2e-capz-azure-disk-windows
+      optional: true
+      run_if_changed: azure.*\.go
+    - name: pull-kubernetes-e2e-capz-azure-file
+      optional: true
+      run_if_changed: azure.*\.go
+    - name: pull-kubernetes-e2e-capz-azure-file-vmss
+      optional: true
+      run_if_changed: azure.*\.go
+    - name: pull-kubernetes-e2e-capz-azure-file-windows
+      optional: true
+      run_if_changed: azure.*\.go
+    - name: pull-kubernetes-e2e-capz-conformance
+      optional: true
+      run_if_changed: azure.*\.go
+    - name: pull-kubernetes-e2e-capz-windows-1-29
```
all of this seems that must be removed
```
ubernetes$ find . | grep -E "azure.*go"
./test/e2e/framework/providers/azure/azure.go
./test/e2e/framework/providers/azure/doc.go
./staging/src/k8s.io/csi-translation-lib/plugins/azure_file.go
./staging/src/k8s.io/csi-translation-lib/plugins/azure_disk_test.go
./staging/src/k8s.io/csi-translation-lib/plugins/azure_disk.go
./staging/src/k8s.io/csi-translation-lib/plugins/azure_file_test.go
./staging/src/k8s.io/client-go/applyconfigurations/core/v1/azurefilepersistentvolumesource.go
./staging/src/k8s.io/client-go/applyconfigurations/core/v1/azurediskvolumesource.go
./staging/src/k8s.io/client-go/applyconfigurations/core/v1/azurefilevolumesource.go
./staging/src/k8s.io/client-go/plugin/pkg/client/auth/azure/azure_stub.go
```

aojea commented on config/tests/jobs/presubmit-jobs.yaml:
```diff
@@ -0,0 +1,158 @@
+    - name: pull-kubernetes-e2e-gci-gce-ingress
+      optional: true
+      run_if_changed: ^(test/e2e/network/|pkg/apis/networking)
+    - name: pull-kubernetes-e2e-gci-gce-ipvs
+      optional: true
+      run_if_changed: ^(test/e2e/network/|pkg/apis/networking|pkg/.*/ipvs/)
```
this is not maintained and I don't have time for it , so we remove it

aojea replied:
done in https://github.com/kubernetes/test-infra/pull/33965

aojea commented on config/tests/jobs/presubmit-jobs.yaml:
```diff
@@ -0,0 +1,158 @@
+    - name: pull-kubernetes-e2e-kind-alpha-beta-features
+      optional: true
+      run_if_changed: ^pkg/features/
+    - name: pull-kubernetes-e2e-kind-ipvs
+      optional: true
+      run_if_changed: ^(test/e2e/network/|pkg/apis/networking|pkg/proxy/ipvs/)
```
ditto https://github.com/kubernetes/test-infra/pull/33958/files#r1886860432

aojea replied:
done in https://github.com/kubernetes/test-infra/pull/33965

pohly commented:
I suggest we remove or update jobs in separate PRs. I can then rebase this one and we merge it when we are happy with the current state.

BenTheElder commented on config/tests/jobs/README.md:
```diff
@@ -8,4 +8,7 @@ To run via bazel: `bazel test //config/tests/jobs/...`
 To run via go: `go test .`
 
+If tests fail, re-run with the `UPDATE_FIXTURE_DATA=true` env variable
+and include the modified files in the PR which updates the jobs.
```
we should probably have a convenient make target since people will be doing this often, normally we have verify scripts spit out the command people need to run when there's generated data to update

pohly replied:
I added `make update-config-fixtures` plus some other make targets like `update-unit` for symmetry with `make unit` which happens to run the Go test. The Go test itself now points to `make update-config-fixtures` because that is faster when the goal is to update the file.

pohly replied:
I switched to `UPDATE_FIXTURE_DATA=true go test ./config/tests/jobs` instead.

BenTheElder commented on config/tests/jobs/presubmit-jobs.yaml:
```diff
@@ -0,0 +1,158 @@
+# AUTOGENERATED by "UPDATE_FIXTURE_DATA=true go test ./config/tests/jobs". DO NOT EDIT!
```
we should move this into a dedicated sub-folder so we can set a specific list of OWNERS for kubernetes presubmits

pohly replied:
Moved to `config/tests/jobs/policy` together with the test, a README.md and the OWNERS file. I copied that from ./config/testgrids/kubernetes/presubmits/OWNERS with some changes:
- removed emeritus
- added `no_parent_owners: true`
- added `sig/infra` as label

That leaves @BenTheElder, @aojea and @test-infra-oncall as reviewer/approver. I probably should remove the latter?

BenTheElder replied:
>  I probably should remove the latter?

Yes. We should have people with emergency ability to e.g. update the CI images being used when there's a regression, but that should not involve changing which jobs run on PRs without talking to a lead.

We should probably add you instead, and ping each other when there are PRs to this.

BenTheElder commented on config/tests/jobs/jobs_test.go:
```diff
@@ -35,10 +35,13 @@ import (
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/sets"
+	yaml "sigs.k8s.io/yaml/goyaml.v3"
```
+1

BenTheElder replied:
filed https://github.com/kubernetes/test-infra/issues/33971

pohly commented:
Interesting...
```console
$ make go-unit
hack/make-rules/go-test/unit.sh
+ umask 0022
+ mkdir -p /work/gopath/src/k8s.io/test-infra/_output
+ /work/gopath/src/k8s.io/test-infra/_bin/gotestsum --junitfile=/work/gopath/src/k8s.io/test-infra/_output/junit-unit.xml -- ./...
✓  config/prow/cluster/build/boskos-resources (cached)
✓  config/tests/jobs/policy (cached)
✓  config/tests/lint (cached)
✓  config/tests/mergelists (cached)
✓  config/tests/jobs (cached)
∅  def/configmap
...
DONE 6812 tests in 1.006s

$ echo $?
0
```

No failure!

But when I run `go test` manually, there is one:
```console
$ go test ./...
...
--- FAIL: TestPresubmitsKubernetesDashboards (0.00s)
    config_test.go:456: pull-kubernetes-e2e-gce-network-proxy-http-connect: should not be in presubmits-kubernetes-blocking because not actually merge-blocking for kubernetes/kubernetes
FAIL
FAIL	k8s.io/test-infra/config/tests/testgrids	1.280s
...
ok  	k8s.io/test-infra/testgrid/pkg/configurator/prow	(cached)
FAIL
```

Not good!

pohly commented:
> No failure!

My `_bin/gotestsum` might have been too old. I can't reproduce after rebuilding it.

pohly commented:
/remove-lifecycle stale

@BenTheElder: did I address your comments?

BenTheElder commented:
/remove-lifecycle stale
... too many things came up.

BenTheElder commented:
I think everything is addressed, just needs updating to the latest jobs (we've removed more from this list ...)
/test all

pohly commented on config/tests/jobs/presubmit-jobs.yaml:
```diff
@@ -0,0 +1,244 @@
+    - name: pull-kubernetes-e2e-capz-conformance
+      optional: true
+      run_if_changed: azure.*\.go
+    - name: pull-kubernetes-e2e-capz-windows-1-32
```
It's unusual that `pull-kubernetes-e2e-capz-windows` changes its name when forked for a release. Should this be changed?

pohly commented on config/tests/jobs/presubmit-jobs.yaml:
```diff
@@ -0,0 +1,184 @@
+    - name: pull-kubernetes-verify
+
+# Non-blocking and automatically triggered presubmits must be stable and
+# (like all presubmits) must have a corresponding periodic.
```
I am unsure whether all current presubmits below qualify. Shall we just waive them through after a cursory glance (= grandfathered in) or do we have to investigate?

aojea commented on config/tests/jobs/jobs_test.go:
```diff
@@ -1459,3 +1462,191 @@ func TestKubernetesProwJobsShouldNotUseDeprecatedScenarios(t *testing.T) {
+			j.Optional = false
+			expected.OptionallyRunIfChanged = append(expected.OptionallyRunIfChanged, j)
+		default:
+			t.Errorf("ERROR: A job with `run_if_changed` must not be `optional: false` (= required).\nSee https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/testing-strategy.md#presubmit-jobs.\n\n%#v", j)
```
I don't remember that run_if_changed can. not be optional, actually I think is better for people developing features to get signal without blocking the rest

aojea replied:
I see your point now, and definitively now I'm unclear on what is best

 @BenTheElder @liggitt @dims since you are commonly involved on these things, should we enforce to fail close or leave this open to interpretation?

BenTheElder replied:
I'm on the fence, we've had problems with it in the past, but for example in kind I use this with `run_if_not_only_changed` (NOT run_if_changed) so essential tests can skip docs-only PRs, and it works well there. But kind also does not have the PR volume to have batch testing for example.

In general I think `run_if_changed` shouldn't be relied on for merge blocking because it's oh so easy to miss the right inputs, it's best thought of as a way to automate some instances of `/test foo` opt-in testing by the PR approver so they don't _always_ have to decide to trigger additional soft-blocking tests.

`run_if_not_only_changed` OTOH is pretty easy to use safely, by excluding say `.github/` from e2e testing.

BenTheElder commented on config/jobs/kubernetes/sig-api-machinery/sig-api-machinery-config.yaml:
```diff
@@ -79,6 +79,7 @@ presubmits:
   - name: pull-kubernetes-e2e-gce-network-proxy-http-connect
     cluster: k8s-infra-prow-build
     always_run: false
+    optional: true
```
Yeah, that seems reasonable, this has caused too many issues in the past.

pohly commented on config/tests/jobs/presubmit-jobs.yaml:
```diff
@@ -0,0 +1,244 @@
+# AUTOGENERATED by "UPDATE_FIXTURE_DATA=true go test ./config/tests/jobs". DO NOT EDIT!
+
+always_run:
+    - name: pull-kubernetes-cmd
```
> Actually, further it should be an approved release-blocking job, or at the very least release-informing.

Agreed.

This brings me back to pull-kubernetes-e2e-gce: do we actually have *any* corresponding periodic job?

pohly replied:
We had ci-kubernetes-e2e-ubuntu-gce-containerd and after some tweaks it is also really identical.

pohly replied:
But that the name is so different just confirms what I said above: without additional annotations the existence of a corresponding periodic will be impossible to check automatically. Even with a pointer it'll be hard because acceptable differences like `--extract=ci/fast/latest-fast` != `--build=quick` depend on the tooling used inside the job and would have to be tracked in some kind of allow-list.

Anyway, future work...

pohly commented:
/remove lifecycle-rotten

BenTheElder reviewed (commented):
/lgtm
/approve

pohly commented:
/test pull-test-infra-unit-test

https://monitoring-eks.prow.k8s.io/d/96Q8oOOZk/builds?orgId=1&refresh=30s&var-org=kubernetes&var-repo=test-infra&var-job=pull-test-infra-unit-test&var-build=All

pohly commented on config/tests/jobs/policy/presubmit-jobs.yaml:
```diff
@@ -0,0 +1,186 @@
+    - name: pull-kubernetes-conformance-kind-ga-only-parallel
+    - name: pull-kubernetes-dependencies
+    - name: pull-kubernetes-e2e-ec2
+      optional: true
```
Another change that I had lost is showing "optional" for jobs. That's relevant for jobs that always run. I know why pull-kubernetes-linter-hints below always runs and is optional, but why does pull-kubernetes-e2e-ec2 fall into that same category? Are contributors and reviewers required to look at failures in that job?

This may or may not be okay, but we should be aware of the potential confusion, therefore it gets shown here now for review.

pohly commented:
Rebased and updated the presubmit list:

```patch
diff --git a/config/tests/jobs/policy/presubmit-jobs.yaml b/config/tests/jobs/policy/presubmit-jobs.yaml
index 9429cf3217..412469ac9c 100644
--- a/config/tests/jobs/policy/presubmit-jobs.yaml
+++ b/config/tests/jobs/policy/presubmit-jobs.yaml
@@ -48,6 +48,8 @@ optional_and_run_if_changed:
       run_if_changed: /(dra|dynamicresources|resourceclaim|deviceclass|resourceslice|resourceclaimtemplate|dynamic-resource-allocation|pkg/apis/resource|api/resource)/.*\.(go|yaml)
     - name: pull-kubernetes-e2e-autoscaling-hpa-cpu
       run_if_changed: ^(pkg\/controller\/podautoscaler\/|test\/e2e\/autoscaling\/horizontal_pod_autoscaling|test\/e2e\/framework\/autoscaling\/)
+    - name: pull-kubernetes-e2e-autoscaling-hpa-cpu-alpha-beta
+      run_if_changed: ^(pkg\/controller\/podautoscaler\/|test\/e2e\/autoscaling\/horizontal_pod_autoscaling|test\/e2e\/framework\/autoscaling\/)
     - name: pull-kubernetes-e2e-capz-azure-disk
       run_if_changed: azure.*\.go
     - name: pull-kubernetes-e2e-capz-azure-disk-vmss
@@ -170,6 +172,8 @@ optional_and_run_if_changed:
       run_if_changed: /staging/src/k8s.io/component-base/logs/
     - name: pull-kubernetes-local-e2e
       run_if_changed: local-up-cluster
+    - name: pull-kubernetes-node-e2e-containerd-1-7-dra
+      run_if_changed: (/dra/|/dynamicresources/|/resourceclaim/|/deviceclass/|/resourceslice/|/resourceclaimtemplate/|/dynamic-resource-allocation/|/pkg/apis/resource/|/api/resource/|/test/e2e_node/dra_).*\.(go|yaml)
     - name: pull-kubernetes-node-e2e-containerd-2-0-dra
       run_if_changed: (/dra/|/dynamicresources/|/resourceclaim/|/deviceclass/|/resourceslice/|/resourceclaimtemplate/|/dynamic-resource-allocation/|/pkg/apis/resource/|/api/resource/|/test/e2e_node/dra_).*\.(go|yaml)
     - name: pull-kubernetes-node-e2e-containerd-2-0-dra-alpha-beta-features
```
